_posts/2018-05-23-gracefully_clean_up_in_grpc_junit_tests.md

```diff
@@ -0,0 +1,115 @@
+---
+layout: post
+title: Let's gracefully clean up in gRPC JUnit tests
+published: false
+permalink: blog/gracefully_clean_up_in_grpc_junit_tests
+author: Dapeng Zhang
+author-link: https://github.com/dapengzhang0
+company: Google
+company-link: https://www.google.com
+---
+
+As a good programmer, it's important that you remember to always clean up gRPC resources such as client channels, servers, and previously attached Contexts whenever they are no longer needed. This is even true for JUnit tests, because otherwise leaked resource will not only linger in your machine forever in some cases but also interfere with subsequent tests. A not-so-bad case is that subsequent tests can not pass because of the leaked resource in the previous test. The worst case is that some subsequent tests pass but in fact they wouldn't have passed if the previously passed test had not leaked resource.
+
+<!--more-->
+
+So cleanup, cleanup, cleanup...\
+and fail the test if any cleanup is not successful.
+
+A typical example is
+
+```java
+public class MyTest {
+  private Server server;
+  private ManagedChannel channel;
+  ...
+  @After
+  public void tearDown() InterruptedException {
+    // assume channel and server are not null
+    channel.shutdownNow();
+    server.shutdownNow();
+    // fail the test if cleanup is not successful
+    assert channel.awaitTermination(5, TimeUnit.SECONDS) : "channel failed to shutdown";
+    assert server.awaitTermination(5, TimeUnit.SECONDS) : "server failed to shutdown";
+  }
+  ...
+}
+```
+
+or to be more graceful
+
+```java
+public class MyTest {
+  private Server server;
+  private ManagedChannel channel;
+  ...
+  @After
+  public void tearDown() throws InterruptedException {
+    // assume channel and server are not null
+    channel.shutdown();
+    server.shutdown();
+    // fail the test if cannot gracefully shutdown
+    try {
+      assert channel.awaitTermination(5, TimeUnit.SECONDS) : "channel cannot be gracefully shutdown";
+      assert server.awaitTermination(5, TimeUnit.SECONDS) : "server cannot be gracefully shutdown";
+    } finally {
+      channel.shutdownNow();
+      server.shutdownNow();
+    }
+  }
+  ...
+}
+```
+
+But that's tedious. You need to write the shutdown boilerplate by yourself. So gRPC testing library introduced something to make it less tedious.
```

Comment:
However, having to add all this to every test so it shuts down gracefully gives you more work to do, as you need to write the shutdown boilerplate by yourself. Because of this, the gRPC testing library introduced something to make the job less tedious.

Reply:
Done.